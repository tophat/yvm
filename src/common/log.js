const defaultLogger = console.error.bind(console) // eslint-disable-line no-console
const capturableLogger = console.log.bind(console) // eslint-disable-line no-console

function log(...args) {
    defaultLogger(...args)
}

log.capturable = function capturableLog(...args) {
    capturableLogger(...args)
}

module.exports = log
