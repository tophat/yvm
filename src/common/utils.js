const os = require('os')
const path = require('path')
const request = require('request')

const log = require('../common/log')

const yvmPath = process.env.YVM_DIR || path.resolve(os.homedir(), '.yvm')
const versionRootPath = rootPath => path.resolve(rootPath, 'versions')

const getExtractionPath = (version, rootPath) =>
    path.resolve(rootPath, 'versions', `v${version}`)

const stripVersionPrefix = tagName =>
    tagName[0] === 'v' ? tagName.substring(1) : tagName

const printVersions = (list, message) => {
    log(message)
    list.forEach(item => {
        log(`  - ${item}`)
    })
}

const getVersionsFromTags = () => {
    const options = {
        url: 'https://api.github.com/repos/yarnpkg/yarn/tags',
        headers: {
            'User-Agent': 'YVM',
        },
    }

    return new Promise((resolve, reject) => {
        request.get(options, (error, response, body) => {
            if (error || response.statusCode !== 200) {
                reject(error)
            } else {
                const tags = JSON.parse(body)
                const tagNames = tags.map(tag => tag.name)
                const versions = tagNames
                    .map(stripVersionPrefix)
                    .filter(version => version[0] > 0)
                resolve(versions)
            }
        })
    })
}

module.exports = {
    getExtractionPath,
    getVersionsFromTags,
    printVersions,
    stripVersionPrefix,
    versionRootPath,
    yvmPath,
}
