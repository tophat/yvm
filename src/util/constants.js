module.exports = {
    USER_AGENT: 'YVM',
}
