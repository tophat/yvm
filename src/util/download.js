const fs = require('fs')
const request = require('request')

const downloadFile = (url, filePath) => {
    const file = fs.createWriteStream(filePath)

    return new Promise((resolve, reject) => {
        const stream = request.get(url).pipe(file)
        stream.on('finish', () => resolve())
        stream.on('error', err => {
            reject(new Error(err))
        })
    })
}

module.exports = {
    downloadFile,
}
