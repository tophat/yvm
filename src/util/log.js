const chalk = require('chalk')

const defaultLogger = console.error.bind(console) // eslint-disable-line no-console
const capturableLogger = console.log.bind(console) // eslint-disable-line no-console

const log = (...args) => log.default(...args)

log.default = function defaultLog(...args) {
    defaultLogger(...args)
}

log.capturable = function capturableLog(...args) {
    capturableLogger(...args)
}

log.info = function infoLog(...args) {
    if (process.env.YVM_VERBOSE) {
        log(...args)
    }
}

log.error = function errorLog(...args) {
    log(chalk.red(...args))
}

log.success = function successLog(...args) {
    log(chalk.green(...args))
}

log.notice = function noticeLog(...args) {
    log(chalk.yellow(...args))
}

module.exports = log
