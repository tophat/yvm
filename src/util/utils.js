import path from 'path'
import request from 'request'
import memoize from 'lodash/memoize'

import {
    USER_AGENT,
    YARN_DOWNLOAD_URL,
    YARN_RELEASES_API_URL,
} from 'util/constants'
import log from 'util/log'

export const versionRootPath = rootPath => path.resolve(rootPath, 'versions')

export const getExtractionPath = (version, rootPath) =>
    path.resolve(rootPath, 'versions', `v${version}`)

export const stripVersionPrefix = tagName =>
    tagName[0] === 'v' ? tagName.substring(1) : tagName

export const getRequest = memoize(async url => {
    const options = {
        url,
        gzip: true,
        headers: {
            'User-Agent': USER_AGENT,
        },
    }
    return new Promise((resolve, reject) => {
        request.get(options, (error, response, body) => {
            const { statusCode, body: responseBody } = response || {}
            if (error || statusCode !== 200) {
                if (responseBody) {
                    if (error) {
                        log(error)
                    }
                    return reject(responseBody)
                }
                return reject(error)
            }
            return resolve(body)
        })
    })
})

export const getVersionDownloadUrl = version =>
    `${YARN_DOWNLOAD_URL}/${version}/yarn-v${version}.tar.gz`

export const getReleasesFromTags = memoize(async () => {
    return getRequest(YARN_RELEASES_API_URL).then(body => {
        return JSON.parse(body).reduce((accumulator, tag) => {
            const version = stripVersionPrefix(tag.name)
            const [major] = version.split('.')
            return Number(major) > 0
                ? Object.assign(accumulator, { [version]: tag })
                : accumulator
        }, {})
    })
})

export const getVersionsFromTags = memoize(async () => {
    try {
        return Object.keys(await getReleasesFromTags())
    } catch (e) {
        log.error(
            'Unable to retrieve remote versions. Please check your network connection',
        )
        return []
    }
})
