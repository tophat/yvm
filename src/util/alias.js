import fs from 'fs'
import { execSync } from 'child_process'
import path from 'path'

import { memoize } from 'lodash'
import chalk from 'chalk'

import log from 'util/log'
import { yvmPath as defaultYvmPath, getNonYvmYarnPathEntries } from 'util/path'
import { getRequest, getVersionsFromTags } from 'util/utils'
import { YARN_STABLE_VERSION_URL } from 'util/constants'

const filterAliasesByName = (pattern, aliases) => {
    return Object.keys(aliases)
        .filter(name => new RegExp(pattern).test(name))
        .map(name => ({ name, value: aliases[name] }))
}

const getAllDependants = ({ name, aliases }) => {
    const visited = []
    const dependants = [name]
    const aliasNames = Object.keys(aliases)
    while (dependants.length) {
        const aliasName = dependants.shift()
        if (visited.includes(aliasName)) {
            continue
        }
        visited.push(aliasName)
        dependants.push(...aliasNames.filter(d => aliases[d] === aliasName))
    }
    return visited
}

export const STORAGE_FILE = '.aliases'

export const DEFAULT = 'default'
export const LATEST = 'latest'
export const LTS = 'lts' // TODO: resolve long term support version
export const STABLE = 'stable'
export const SYSTEM = 'system'
export const UNRESOLVED = undefined
export const NOT_AVAILABLE = 'N/A'
export const DEFAULT_NAMES = [LATEST, STABLE, SYSTEM]
export const RESERVED_NAMES = [...DEFAULT_NAMES, LTS, NOT_AVAILABLE]

export const resolveLatest = memoize(
    async () => (await getVersionsFromTags())[0],
)

export const resolveStable = memoize(async () => {
    const version = await getRequest(YARN_STABLE_VERSION_URL)
    return (version && version.trim()) || UNRESOLVED
})

export const resolveSystem = memoize(async ({ shell } = {}) => {
    for (const pathToExec of getNonYvmYarnPathEntries(shell)) {
        try {
            const execCommand = `${pathToExec} --version 2> /dev/null`
            return String(execSync(execCommand)).trim()
        } catch (error) {
            log.info(error.message)
        }
    }
    return NOT_AVAILABLE
})

export const isReserved = name => RESERVED_NAMES.includes(name)

export const resolveReserved = memoize(async (name, args = {}) => {
    const resolver =
        {
            [LATEST]: resolveLatest,
            [STABLE]: resolveStable,
            [SYSTEM]: resolveSystem,
        }[name] || (async () => NOT_AVAILABLE)
    return resolver(args)
})

export const getDefaultAliases = () => {
    return DEFAULT_NAMES.reduce(
        (aliases, name) => Object.assign(aliases, { [name]: UNRESOLVED }),
        {},
    )
}

export const getUserAliases = memoize(async (yvmPath = defaultYvmPath) => {
    const aliases = { [DEFAULT]: STABLE }
    const aliasStoragePath = path.join(yvmPath, STORAGE_FILE)
    try {
        const aliasesJSON = fs.readFileSync(aliasStoragePath, 'utf8')
        Object.assign(aliases, JSON.parse(aliasesJSON))
    } catch (e) {
        log.info(e.message)
    }
    return aliases
})

export const getMatchingAliases = async (pattern, yvmPath = defaultYvmPath) => {
    return filterAliasesByName(pattern, await getUserAliases(yvmPath))
}

export const setAlias = async ({ name, version, yvmPath = defaultYvmPath }) => {
    const aliasStoragePath = path.join(yvmPath, STORAGE_FILE)
    try {
        if (isReserved(name)) {
            throw new Error(`'${name}' is reserved. Please choose another.`)
        }
        const aliases = await getUserAliases(yvmPath)
        aliases[name] = version
        fs.writeFileSync(aliasStoragePath, JSON.stringify(aliases))
        return true
    } catch (e) {
        log(`Unable to set alias. ${e.message}`)
        log.info(e.stack)
        return false
    }
}

export const unsetAlias = async ({
    name,
    force = false,
    recursive = false,
    yvmPath = defaultYvmPath,
}) => {
    const aliasStoragePath = path.join(yvmPath, STORAGE_FILE)
    try {
        if (isReserved(name)) {
            throw new Error(`'${name}' is reserved by yvm.`)
        }
        let deleted = false
        const aliases = await getUserAliases(yvmPath)
        const aliasNameList = Object.keys(aliases)
        const dependants = aliasNameList.filter(d => aliases[d] === name)
        if (dependants.length && !force && !recursive) {
            log.notice(`The following aliases will be orphaned: ${dependants}
Rerun with '--force' to remove just this alias
Or with '--recursive' to remove all dependants`)
        } else {
            const namesToRemove = recursive
                ? getAllDependants({ name, aliases })
                : [name]
            const newAliases = {}
            for (const aliasName of aliasNameList) {
                if (namesToRemove.includes(aliasName)) continue
                newAliases[aliasName] = aliases[aliasName]
            }
            deleted = aliases[name] !== newAliases[name]
            log.info(`The following aliases were removed: ${namesToRemove}`)
            fs.writeFileSync(aliasStoragePath, JSON.stringify(newAliases))
        }
        if (dependants.length && deleted && !recursive) {
            log.notice(`The following aliases are now orphaned: ${dependants}`)
        }
        return deleted
    } catch (e) {
        log(`Unable to unset ${e.message}`)
        log.info(e.stack)
        return false
    }
}

export const resolveAlias = memoize(
    async ({
        versionString,
        currentAliases,
        visitedAliases = [],
        yvmPath = defaultYvmPath,
    }) => {
        const chain = visitedAliases.join(', ')
        const resolveErrorMessage = msg => `${msg}: '${versionString}' ${chain}`
        if (visitedAliases.includes(versionString)) {
            throw new Error(resolveErrorMessage`Cyclic chain detected`)
        }
        const aliases = currentAliases || (await getUserAliases(yvmPath))
        if (versionString in aliases && !isReserved(versionString)) {
            return await resolveAlias({
                versionString: aliases[versionString],
                currentAliases: aliases,
                visitedAliases: [...visitedAliases, versionString],
                yvmPath,
            })
        }
        return {
            visited: visitedAliases,
            version: versionString,
        }
    },
)

export const resolveAliases = memoize(async (yvmPath = defaultYvmPath) => {
    const currentAliases = {
        ...(await getUserAliases(yvmPath)),
        ...getDefaultAliases(),
    }
    const resolvedAliases = {}
    for (const versionString of Object.keys(currentAliases)) {
        if (versionString in resolvedAliases) {
            continue
        }
        const { version, visited } = await resolveAlias({
            versionString,
            currentAliases,
            yvmPath,
        }).catch(() => ({ version: undefined, visited: [] }))
        for (const alias of [versionString, ...visited]) {
            resolvedAliases[alias] = {
                value: currentAliases[alias],
                version,
            }
        }
    }
    return resolvedAliases
})

export const resolveMatchingAliases = memoize(
    async (pattern = '', yvmPath = defaultYvmPath) => {
        return filterAliasesByName(pattern, await resolveAliases(yvmPath))
    },
)

export const getFormatter = (
    allVersions,
    installedVersions,
    currentVersion,
) => (name, version, target) => {
    const defaultStyler = chalk.grey
    let termStyler = defaultStyler
    const isAvailable = allVersions.includes(target)
    if (target === currentVersion) {
        termStyler = chalk.green
    } else if (installedVersions.includes(target)) {
        termStyler = chalk.yellow
    } else if (isAvailable) {
        termStyler = chalk.white
    } else {
        termStyler = chalk.red
    }
    if (isReserved(name)) {
        termStyler = termStyler.bold
    }
    const parts = [termStyler(name), ' â†’ ']
    const targetVersion = termStyler(isAvailable ? target : NOT_AVAILABLE)
    if (version) {
        parts.push(chalk.white(version), ' (', targetVersion, ')')
    } else {
        parts.push(targetVersion)
    }

    return defaultStyler(parts.join(''))
}
