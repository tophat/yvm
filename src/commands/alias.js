import log from '../util/log'
import { getFormatter, resolveMatchingAliases, setAlias } from '../util/alias'
import {
    getVersionInUse,
    getYarnVersions,
    resolveVersion,
} from '../util/version'
import { getVersionsFromTags } from '../util/utils'

export const alias = async (nameOrPattern, maybeVersion) => {
    try {
        const format = getFormatter(
            await getVersionsFromTags(),
            getYarnVersions(),
            await getVersionInUse(),
        )

        const safeResolveVersion = async versionString =>
            resolveVersion({ versionString }).catch(e => log.info(e.message))

        if (typeof maybeVersion === 'string') {
            const name = nameOrPattern
            const version = maybeVersion
            let targetVersion
            if (await setAlias({ name, version })) {
                targetVersion = await safeResolveVersion(version)
            }
            const message = format(name, version, targetVersion)
            if (targetVersion) {
                log(message)
                return 0
            } else {
                log.error(message)
                return 1
            }
        } else {
            const pattern = nameOrPattern || ''
            const matchingAliases = await resolveMatchingAliases(pattern)
            for (const {
                name,
                value: { value, version: versionString },
            } of matchingAliases) {
                const version = await safeResolveVersion(versionString)
                log(format(name, value, version))
            }
            const noMatchFound =
                pattern.length > 0 && matchingAliases.length < 1
            return noMatchFound ? 1 : 0
        }
    } catch (e) {
        log.error(e.message)
        return 2
    }
}
