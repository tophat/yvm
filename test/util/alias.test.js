import childProcess from 'child_process'

import * as mockProps from 'jest-mock-props'
import { fs, vol } from 'memfs'

mockProps.extend(jest)

import log from 'util/log'
import * as utils from 'util/utils'
import * as alias from 'util/alias'

jest.spyOn(childProcess, 'execSync')
const pathSpy = jest.spyOnProp(process.env, 'PATH')
const fishUserPathSpy = jest.spyOnProp(process.env, 'fish_user_paths')

jest.spyOn(utils, 'getRequest')
jest.spyOn(utils, 'getVersionsFromTags')

describe('alias util', () => {
    const yvmPath = '/Users/tophat/.yvm'
    jest.spyOn(log, 'info')

    afterAll(jest.restoreAllMocks)

    describe('resolveLatest', () => {
        it('gets latest version', async () => {
            utils.getVersionsFromTags.mockReturnValueOnce([
                '1.4.0',
                '1.3.0',
                '1.2.0',
            ])
            expect(await alias.resolveLatest()).toBe('1.4.0')
        })
    })

    describe('resolveStable', () => {
        afterEach(() => {
            alias.resolveStable.cache.clear()
        })

        it('gets stable version', async () => {
            utils.getRequest.mockReturnValueOnce(' 1.15.2 ')
            expect(await alias.resolveStable()).toBe('1.15.2')
            expect(utils.getRequest).toHaveBeenCalledWith(
                'https://yarnpkg.com/latest-version',
            )
        })

        it('does not get stable version', async () => {
            utils.getRequest.mockReturnValueOnce('  ')
            expect(await alias.resolveStable()).toBe(alias.UNRESOLVED)
        })
    })

    describe('resolveSystem', () => {
        beforeEach(() => {
            vol.fromJSON({
                '/usr/local/bin/yarn': 'mock exec',
            })
        })

        afterEach(() => {
            alias.resolveSystem.cache.clear()
            pathSpy.mockReset()
            fishUserPathSpy.mockReset()
        })

        afterAll(() => {
            vol.reset()
        })

        it('returns nothing if yarn executable fails to run', async () => {
            const mockError = new Error('not executable')
            childProcess.execSync.mockImplementationOnce(() => {
                throw mockError
            })
            expect(await alias.resolveSystem()).toEqual(alias.NOT_AVAILABLE)
            expect(log.info).toHaveBeenCalledWith(mockError.message)
        })

        it('gets first non yvm yarn executable', async () => {
            pathSpy.mockValue(`${yvmPath}/versions/v1.13.0/bin:/usr/local/bin:`)
            childProcess.execSync.mockReturnValueOnce('1.14.0')
            expect(await alias.resolveSystem()).toEqual('1.14.0')
            expect(childProcess.execSync).toHaveBeenCalledWith(
                expect.stringContaining('usr/local/bin/yarn'),
            )
        })

        it('returns nothing if yarn not found in path', async () => {
            pathSpy.mockValue('/Users/tophat/.nvm/versions/node/v6.11.5/bin:')
            expect(await alias.resolveSystem()).toEqual(alias.NOT_AVAILABLE)
        })

        it('gets first non yvm yarn executable in fish shell', async () => {
            fishUserPathSpy.mockValue(
                `${yvmPath}/versions/v1.13.0/bin /usr/local/bin`,
            )
            childProcess.execSync.mockReturnValueOnce('1.14.0')
            expect(await alias.resolveSystem({ shell: 'fish' })).toEqual(
                '1.14.0',
            )
            expect(childProcess.execSync).toHaveBeenCalledWith(
                expect.stringContaining('usr/local/bin/yarn'),
            )
        })

        it('returns nothing if yarn not found in fish user path', async () => {
            fishUserPathSpy.mockValue(
                '/Users/tophat/.nvm/versions/node/v6.11.5/bin ',
            )
            expect(await alias.resolveSystem({ shell: 'fish' })).toEqual(
                alias.NOT_AVAILABLE,
            )
        })
    })

    describe('getDefaultAliases', () => {
        it('matches snapshot', () => {
            expect(alias.getDefaultAliases()).toMatchObject({
                [alias.LATEST]: alias.UNRESOLVED,
                [alias.STABLE]: alias.UNRESOLVED,
                [alias.SYSTEM]: alias.UNRESOLVED,
            })
        })
    })

    describe('getUserAliases', () => {
        const aliasesPath = `${yvmPath}/${alias.STORAGE_FILE}`
        beforeEach(() => {
            vol.fromJSON({
                [aliasesPath]: '',
            })
            alias.getUserAliases.cache.clear()
        })

        afterAll(() => {
            vol.reset()
        })

        it('sets default to stable', async () => {
            expect(await alias.getUserAliases(yvmPath)).toEqual({
                [alias.DEFAULT]: alias.STABLE,
            })
        })

        it('overwrites default with loaded value', async () => {
            const mockDefaultAlias = 'someothervalue'
            vol.fromJSON({
                [aliasesPath]: `{"default":"${mockDefaultAlias}"}`,
            })
            expect(await alias.getUserAliases(yvmPath)).toEqual({
                [alias.DEFAULT]: mockDefaultAlias,
            })
        })

        it('uses default when aliases file is corrupted', async () => {
            vol.fromJSON({
                [aliasesPath]: '"default":"testing"',
            })
            expect(await alias.getUserAliases(yvmPath)).toEqual({
                [alias.DEFAULT]: alias.STABLE,
            })
        })

        describe('getMatchingAliases', () => {
            it('returns all aliases with empty pattern', async () => {
                const mockAliases = {
                    default: '1.15.0',
                    other: '1.2.0',
                }
                const expectedAliases = [
                    { name: 'default', value: '1.15.0' },
                    { name: 'other', value: '1.2.0' },
                ]
                vol.fromJSON({
                    [aliasesPath]: JSON.stringify(mockAliases),
                })
                expect(
                    await alias.getMatchingAliases(undefined, yvmPath),
                ).toEqual(expectedAliases)
            })

            it('returns only matching aliases', async () => {
                const mockAliases = {
                    default: '1.15.0',
                    other: '1.2.0',
                }
                const expectedAliases = [{ name: 'other', value: '1.2.0' }]
                vol.fromJSON({
                    [aliasesPath]: JSON.stringify(mockAliases),
                })
                expect(await alias.getMatchingAliases('oth', yvmPath)).toEqual(
                    expectedAliases,
                )
            })
        })
    })

    describe('setAlias', () => {
        const aliasesPath = `${yvmPath}/${alias.STORAGE_FILE}`
        jest.spyOn(log, 'default')
        beforeEach(() => {
            vol.fromJSON({
                [aliasesPath]: '',
            })
            alias.getUserAliases.cache.clear()
        })

        afterAll(() => {
            vol.reset()
        })

        it('sets user alias', async () => {
            expect(
                await alias.setAlias({
                    name: 'test',
                    version: '1.7.0',
                    yvmPath,
                }),
            ).toBe(true)
            expect(await alias.getUserAliases(yvmPath)).toEqual({
                [alias.DEFAULT]: alias.STABLE,
                test: '1.7.0',
            })
        })

        it.each(alias.RESERVED_NAMES.map(name => [name]))(
            'does not override reserved alias %s',
            async () => {
                expect(
                    await alias.setAlias({
                        name: alias.STABLE,
                        version: '1.7.0',
                        yvmPath,
                    }),
                ).toBe(false)
                expect(log.default).toHaveBeenCalledWith(
                    expect.stringContaining('Unable to set alias'),
                )
            },
        )
    })

    describe('unsetAlias', () => {
        const aliasesPath = `${yvmPath}/${alias.STORAGE_FILE}`
        jest.spyOn(fs, 'writeFileSync')
        jest.spyOn(log, 'default')

        beforeEach(() => {
            vol.fromJSON({
                [aliasesPath]: '',
            })
        })

        afterEach(() => {
            jest.clearAllMocks()
            alias.getUserAliases.cache.clear()
        })

        afterAll(() => {
            vol.reset()
        })

        it('does not unset reserved alias', async () => {
            expect(
                await alias.unsetAlias({
                    name: alias.STABLE,
                    yvmPath,
                }),
            ).toBe(false)
            expect(log.default).toHaveBeenCalledWith(
                expect.stringContaining('Unable to unset'),
            )
        })

        it('does not unset aliases with dependants', async () => {
            vol.fromJSON({
                [aliasesPath]: JSON.stringify({
                    one: 'zero',
                    two: 'one',
                    zero: '1.7.0',
                }),
            })
            await alias.getUserAliases(yvmPath)
            expect(
                await alias.unsetAlias({
                    name: 'zero',
                    yvmPath,
                }),
            ).toBe(false)
            expect(fs.writeFileSync).not.toHaveBeenCalled()
            expect(log.default).toHaveBeenCalledWith(
                expect.stringContaining(
                    'The following aliases will be orphaned',
                ),
            )
        })

        it('unset aliases with dependants with the force option', async () => {
            vol.fromJSON({
                [aliasesPath]: JSON.stringify({
                    one: 'zero',
                    two: 'one',
                    zero: '1.7.0',
                }),
            })
            const mockAliases = await alias.getUserAliases(yvmPath)
            expect(
                await alias.unsetAlias({
                    name: 'zero',
                    force: true,
                    yvmPath,
                }),
            ).toBe(true)
            delete mockAliases.zero
            expect(fs.writeFileSync).toHaveBeenCalledWith(
                aliasesPath,
                JSON.stringify(mockAliases),
            )
        })

        it('unset all dependant aliases with the recursive option', async () => {
            vol.fromJSON({
                [aliasesPath]: JSON.stringify({
                    one: 'zero',
                    two: 'one',
                    zero: '1.7.0',
                }),
            })
            await alias.getUserAliases(yvmPath)
            expect(
                await alias.unsetAlias({
                    name: 'zero',
                    recursive: true,
                    yvmPath,
                }),
            ).toBe(true)
            expect(fs.writeFileSync).toHaveBeenCalledWith(
                aliasesPath,
                JSON.stringify({ [alias.DEFAULT]: alias.STABLE }),
            )
        })

        it('does not crash on cyclic alias dependancy', async () => {
            vol.fromJSON({
                [aliasesPath]: JSON.stringify({
                    one: 'zero',
                    two: 'one',
                    zero: 'two',
                }),
            })
            expect(
                await alias.unsetAlias({
                    name: 'zero',
                    recursive: true,
                    yvmPath,
                }),
            ).toBe(true)
            expect(fs.writeFileSync).toHaveBeenCalledWith(
                aliasesPath,
                JSON.stringify({ [alias.DEFAULT]: alias.STABLE }),
            )
        })

        it('unsets alias successfully', async () => {
            vol.fromJSON({
                [aliasesPath]: JSON.stringify({
                    one: 'zero',
                    two: 'one',
                    zero: '1.7.0',
                }),
            })
            const mockAliases = await alias.getUserAliases(yvmPath)
            expect(
                await alias.unsetAlias({
                    name: 'two',
                    force: true,
                    yvmPath,
                }),
            ).toBe(true)
            delete mockAliases.two
            expect(fs.writeFileSync).toHaveBeenCalledWith(
                aliasesPath,
                JSON.stringify(mockAliases),
            )
        })
    })

    describe('resolveAliases', () => {
        const aliasesPath = `${yvmPath}/${alias.STORAGE_FILE}`
        jest.spyOn(fs, 'writeFileSync')
        jest.spyOn(log, 'default')

        beforeEach(() => {
            vol.fromJSON({
                [aliasesPath]: '',
            })
        })

        afterEach(() => {
            vol.reset()
            jest.clearAllMocks()
            alias.getUserAliases.cache.clear()
            alias.resolveAliases.cache.clear()
            alias.resolveMatchingAliases.cache.clear()
        })

        it('matches all aliases on empty pattern', async () => {
            const version = '1.8.0'
            vol.fromJSON({
                [aliasesPath]: JSON.stringify({
                    one: 'zero',
                    two: 'one',
                    zero: version,
                }),
            })
            const result = await alias.resolveMatchingAliases(
                undefined,
                yvmPath,
            )
            expect(result).toMatchSnapshot()
        })

        it('does not crash on cyclic alias dependancy', async () => {
            vol.fromJSON({
                [aliasesPath]: JSON.stringify({
                    one: 'zero',
                    two: 'one',
                    zero: 'two',
                }),
            })
            const result = await alias.resolveMatchingAliases(
                undefined,
                yvmPath,
            )
            expect(result).toMatchSnapshot()
        })

        it('resolves only matching aliases', async () => {
            const version = '1.8.0'
            vol.fromJSON({
                [aliasesPath]: JSON.stringify({
                    one: 'zero',
                    two: 'one',
                    zero: version,
                }),
            })
            expect(await alias.resolveMatchingAliases('one', yvmPath)).toEqual([
                { name: 'one', value: { value: 'zero', version } },
            ])
        })

        it('returns not available when unable to resolve reserved', async () => {
            expect(await alias.resolveReserved('somealias')).toEqual(
                alias.NOT_AVAILABLE,
            )
        })
    })

    describe('getFormatter', () => {
        const currentVersion = '1.13.0'
        const installedVersions = [currentVersion, '1.8.0', '1.3.0']
        const allVersions = ['1.2.0', '1.5.0', ...installedVersions]
        const formatter = alias.getFormatter(
            allVersions,
            installedVersions,
            currentVersion,
        )
        const inputCases = [
            ['current', '^1.1', '1.13.0'],
            ['installed', '1.3', '1.3.0'],
            ['available', '1.5', '1.5.0'],
            ['unavailable', '1.1', '1.1.0'],
            ['undefined', '1.2.0', undefined],
            [alias.STABLE, undefined, '1.13.0'],
            [alias.SYSTEM, undefined, '1.7.0'],
        ]
        it.each(inputCases)(
            'formats and colors inputs correctly %s alias',
            (name, version, target) => {
                const formatted = formatter(name, version, target)
                log(formatted)
                expect(formatted).toMatchSnapshot()
            },
        )
    })
})
