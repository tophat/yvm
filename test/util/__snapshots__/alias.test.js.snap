// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`alias getFormatter formats and colors inputs correctly available alias 1`] = `"[90m[37mavailable[90m â†’ [37m1.5[90m ([37m1.5.0[90m)[39m"`;

exports[`alias getFormatter formats and colors inputs correctly current alias 1`] = `"[90m[32mcurrent[90m â†’ [37m^1.1[90m ([32m1.13.0[90m)[39m"`;

exports[`alias getFormatter formats and colors inputs correctly installed alias 1`] = `"[90m[33minstalled[90m â†’ [37m1.3[90m ([33m1.3.0[90m)[39m"`;

exports[`alias getFormatter formats and colors inputs correctly stable alias 1`] = `"[90m[32m[1mstable[22m[90m â†’ [32m[1m1.13.0[22m[90m[39m"`;

exports[`alias getFormatter formats and colors inputs correctly system alias 1`] = `"[90m[31m[1msystem[22m[90m â†’ [31m[1mN/A[22m[90m[39m"`;

exports[`alias getFormatter formats and colors inputs correctly unavailable alias 1`] = `"[90m[31munavailable[90m â†’ [37m1.1[90m ([31mN/A[90m)[39m"`;

exports[`alias getFormatter formats and colors inputs correctly undefined alias 1`] = `"[90m[31mundefined[90m â†’ [37m1.2.0[90m ([31mN/A[90m)[39m"`;

exports[`alias resolveAliases does not crash on cyclic alias dependancy 1`] = `
Array [
  Object {
    "name": "default",
    "value": Object {
      "value": "stable",
      "version": "stable",
    },
  },
  Object {
    "name": "one",
    "value": Object {
      "value": "zero",
      "version": undefined,
    },
  },
  Object {
    "name": "two",
    "value": Object {
      "value": "one",
      "version": undefined,
    },
  },
  Object {
    "name": "zero",
    "value": Object {
      "value": "two",
      "version": undefined,
    },
  },
  Object {
    "name": "latest",
    "value": Object {
      "value": undefined,
      "version": "latest",
    },
  },
  Object {
    "name": "stable",
    "value": Object {
      "value": undefined,
      "version": "stable",
    },
  },
  Object {
    "name": "system",
    "value": Object {
      "value": undefined,
      "version": "system",
    },
  },
]
`;

exports[`alias resolveAliases matches all aliases on empty pattern 1`] = `
Array [
  Object {
    "name": "default",
    "value": Object {
      "value": "stable",
      "version": "stable",
    },
  },
  Object {
    "name": "one",
    "value": Object {
      "value": "zero",
      "version": "1.8.0",
    },
  },
  Object {
    "name": "zero",
    "value": Object {
      "value": "1.8.0",
      "version": "1.8.0",
    },
  },
  Object {
    "name": "two",
    "value": Object {
      "value": "one",
      "version": "1.8.0",
    },
  },
  Object {
    "name": "latest",
    "value": Object {
      "value": undefined,
      "version": "latest",
    },
  },
  Object {
    "name": "stable",
    "value": Object {
      "value": undefined,
      "version": "stable",
    },
  },
  Object {
    "name": "system",
    "value": Object {
      "value": undefined,
      "version": "system",
    },
  },
]
`;
